src/connections.jl

```diff
@@ -193,14 +199,19 @@ end
 
 # AbstractLock primitives:
 # https://github.com/JuliaLang/julia/blob/master/base/condition.jl#L18
-Base.lock(conn::Connection) = lock(conn.lock)
-Base.unlock(conn::Connection) = unlock(conn.lock)
-Base.trylock(conn::Connection) = trylock(conn.lock)
-Base.islocked(conn::Connection) = islocked(conn.lock)
+Base.lock(conn::Connection) = acquire(conn.semaphore)
+Base.unlock(conn::Connection) = release(conn.semaphore)
+Base.islocked(conn::Connection) = conn.semaphore.curr_cnt >= conn.semaphore.sem_size
 
-# AbstractLock conventions:
-Base.lock(f, conn::Connection) = lock(f, conn.lock)
-Base.trylock(f, conn::Connection) = trylock(f, conn.lock)
+# AbstractLock convention:
+function Base.lock(f, conn::Connection)
+    lock(conn)
+    try
+        return f()
+    finally
+        unlock(conn)
+    end
+end
```

Comment:
why did yo unroll this rather than using `lock(f, conn.lock)`?
Is it because Semaphore's don't implement that pattern?
if so we should open an issue about it

Reply:
Yes

Reply:
Also, `lock(f, the_lock)` is restricted to `::AbstractLock`, even though It probably shouldn't be. I mentioned that one a draft PR I made that no one looked at.

Reply:
https://github.com/JuliaLang/julia/pull/32299